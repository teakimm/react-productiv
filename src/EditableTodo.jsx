import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";
import TopTodo from "./TopTodo";

/** Show editable todo item.
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * State:
 * -showEditForm(true or false)
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove, complete }) {

  const [showEditForm, setShowEditForm] = useState(false);


  /** Toggle if this is being edited */
  function toggleEdit() {
    setShowEditForm(showEditForm => showEditForm = !showEditForm);
  }

  /** Call remove fn passed to this. */
  function handleDelete() {
    remove(todo.id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave(formData) {
    update({ id: todo.id, ...formData });
    toggleEdit();
  }

  function completedOrNot() {
    complete(todo);
  }

  return (



    <div className="EditableTodo">

      {showEditForm
        ? <TodoForm initialFormData={todo} handleSave={handleSave} />
        :
        <div className="mb-3">
          <div className="float-end text-sm-end">
            <button
              className="EditableTodo-toggle btn-link btn btn-sm"
              onClick={toggleEdit}>
              Edit
            </button>
            <button
              className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
              onClick={handleDelete}>
              Del
            </button>
            <button
              className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
              onClick={completedOrNot}>
              Completed
            </button>
          </div>
          <Todo
            todo={todo}
          />
        </div>
      }
    </div>
  );
}

export default EditableTodo;
