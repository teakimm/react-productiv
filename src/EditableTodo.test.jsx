import { describe, it, expect, test } from "vitest";

import React from "react";
import { fireEvent, render } from "@testing-library/react";
import EditableTodo from "./EditableTodo";

const testTodo = {
    id: "anyId",
    title: "testTitle",
    description: "hello world",
    priority: 3
};


it("renders without crashing", function () {
    render(<EditableTodo todo={testTodo} />);
});

it("matches snapshot", function () {
    const { container } = render(<EditableTodo todo={testTodo} />);
    expect(container).toMatchSnapshot();
});

it("contains todo", function () {
    const result = render(<EditableTodo todo={testTodo} />);


    expect(
        result.queryByText("testTitle")
    ).toBeInTheDocument();

    expect(
        result.queryByText("hello world")
    ).toBeInTheDocument();

    expect(
        result.queryByText("not a description")
    ).not.toBeInTheDocument();
});

it("contains edit and delete button", function () {
    const { container } = render(<EditableTodo todo={testTodo} />);


    expect(
        container.querySelector(".EditableTodo-toggle")
    ).toBeInTheDocument();

    expect(
        container.querySelector(".EditableTodo-delBtn")
    ).toBeInTheDocument();

});


it("test update", function () {
    const { container, debug } = render(<EditableTodo todo={testTodo} />);
    const updateButton = container.querySelector(".EditableTodo-toggle");

    fireEvent.click(updateButton);

    expect(
        container.querySelector(".NewTodoForm")
    ).toBeInTheDocument();

    expect(
        container.querySelector("#newTodo-description")
    ).toBeInTheDocument();

    expect(
        container.querySelector("#newTodo-title")
    ).toBeInTheDocument();

    expect(
        container.querySelector("#newTodo-priority")
    ).toBeInTheDocument();
});