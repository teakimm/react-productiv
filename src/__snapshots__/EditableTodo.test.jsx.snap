// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`contains edit and delete buttons 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            testTitle
          </b>
           
          <small>
            (
            priority: 3
            )
          </small>
        </div>
        <div>
          <small>
            hello world
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            testTitle
          </b>
           
          <small>
            (
            priority: 3
            )
          </small>
        </div>
        <div>
          <small>
            hello world
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
